### ISSUE TITLE
Move the arms with the Valve knuckles

### ISSUE THREAD

**Issue by Miouyouyou** [GodotVR/godot_openxr#11]
Greetings,

This is more a "I have no real clue of what I'm doing, so I'm asking for pointers", rather than a real bug report.

So, I'm just starting to fiddle with OpenXR (and Godot) on Linux, and starting to piece out all the required libraries and tools together...

I'm currently using Arch Linux, and I'm using AUR **git** packages for many of the required libraries.
I've installed openxr, monado and libsurvive and everything seems to more or less working. **xrgears** do not show any arm, but I don't know if it can.

So, I compiled and tried this example but I can't seem to move any arm using the Valve Index Knuckles controllers. I can see the scene and move the head inside the Valve headset, look at the wonderful table, but that's it.
No arms and no lighthouse detection.

I added a few "fprintf" to **OpenXRApi::suggestActions** and the logs seem to suggest that the OpenXR is actually trying to use the controllers. But it's either not receiving the inputs from the actual hardware, or something is going wrong with the Godot setup. I'm pretty sure it's the former, though.

```
OpenXR: initialising OpenXR context
OpenXR Runtime only supports OpenGL version 4.5 - 4.6!
Graphics: Display 0x6594050, Context 50af480, Drawable a600005
Using OpenGL version: 4.6 (Core Profile) Mesa 20.1.7
Using OpenGL Renderer: Radeon RX Vega (VEGA10, DRM 3.38.0, 5.8.8-zen1-1-zen, LLVM 10.0.1)
Swapchain Formats
8C43
Using SRGB swapchain!
8059
881A
8058
81A5
8CAC
8CAD
Yay ! We got /interaction_profiles/khr/simple_controller !
Suggesting actions for /interaction_profiles/khr/simple_controller, 4 bindings
Binding 0, Action 0 => 0
Binding 1, Action 0 => 1
Binding 2, Action 1 => 0
Binding 3, Action 1 => 1
Simple bindings OK ?
YES !
Yay ! We got /interaction_profiles/valve/index_controller !
Suggesting actions for /interaction_profiles/valve/index_controller, 8 bindings
Binding 0, Action 0 => 0
Binding 1, Action 0 => 1
Binding 2, Action 1 => 0
Binding 3, Action 1 => 1
Binding 4, Action 2 => 0
Binding 5, Action 2 => 1
Binding 6, Action 3 => 0
Binding 7, Action 3 => 1
Simple bindings OK ?
YES !
Yay ! We got /interaction_profiles/mndx/ball_on_a_stick_controller !
Suggesting actions for /interaction_profiles/mndx/ball_on_a_stick_controller, 8 bindings
Binding 0, Action 0 => 0
Binding 1, Action 0 => 1
Binding 2, Action 1 => 0
Binding 3, Action 1 => 1
Binding 4, Action 2 => 0
Binding 5, Action 2 => 1
Binding 6, Action 3 => 0
Binding 7, Action 3 => 1
Simple bindings OK ?
YES !
initialized controllers 1 2
OpenXR init succeeded
AR VR Interface initialized
Started ARVR
EVENT: session state changed to 1
EVENT: session state changed to 2
EVENT: interaction profile changed!
Event: Interaction profile changed for /user/hand/left: /interaction_profiles/valve/index_controller
Event: Interaction profile changed for /user/hand/right: /interaction_profiles/valve/index_controller
EVENT: interaction profile changed!
Event: Interaction profile changed for /user/hand/left: /interaction_profiles/valve/index_controller
Event: Interaction profile changed for /user/hand/right: /interaction_profiles/valve/index_controller
```

Now, I'm a bit a lost when it comes to debug the whole situation.

I installed **libsurvive** and **survive-buttons** is able to detect the knuckles input but, after that, I don't know which software I could use to assert that everything is setup correctly, software wise.
**xrgears** show up the gears inside the VR Headset, but again no knuckles detection it seems, so I guess that something is going wrong between OpenXR and libsurvive.

Just for information, I'm able to use SteamVR and play VR titles on Steam like SteamVR Environments or VRChat (through Proton) just fine.
However, I didn't execute SteamVR, nor **vrmonitor** while playing with this demo, as they seem to badly interfere with **monado-service**.

Anyway, there's two questions here :

1. Is this example supposed to show moving arms when using the Valve Index controllers ? I guess so, since the scene has 2 "ARVRController" with meshes attached... I tried various functions to check if Godot knows what device is behind which ARVRController, but even things like 
```javascript
	var controller = $ARVROrigin/ARVRController
	print(controller.get_controller_name())
```
just show **lefthand** (for the first controller)...
2. Do you have a checklist to determined if the software setup is alright ? Or some ideas of softwares I could test before ?

**Comment by Miouyouyou** [comment]
Alright, compiling **monado** after explicitly enabling **-DXRT_BUILD_DRIVER_SURVIVE** , I was able to get Knuckles and Lighthouse support working with either **hello_xr** and the Godot example.

So I can move my arms inside the Godot example, now.

Now, seems the libsurvive header provides the implementation through "static inline" exports, it seems there's no way to check whether **monado** is compiled with libsurvive, beside testing it directly.
`ldd /usr/lib/libopenxr_monado.so | grep -i surv` returns nothing, and so does `readelf -a /usr/lib/libopenxr_monado.so | grep -i surv`.

Anyway, it's working now. I'll now play with the actual examples provided by Godot.

**Comment by ChristophHaag** [comment]
xrgears has no controller input either way btw. hello_xr from openxr-sdk (aur: openxr-utils-git) or [our openxr-simple-example](https://gitlab.freedesktop.org/monado/demos/openxr-simple-example) would be reasonably small applications for testing.

Monado indeed doesn't have a way to list built/available drivers or select them. It's in the "nice to have" category of the roadmap.

Generally you get some survive related output on stdout when it is used, but there's also `monado-cli probe` showing a string description of the devices that are found, where the ones from the survive driver are prefixed with `Survive`.

**Comment by Miouyouyou** [comment]
>  there's also monado-cli probe showing a string description of the devices that are found, where the ones from the survive driver are prefixed with Survive

Aaaah, I didn't know that one.

That said, once I got the whole setup, things started to click in my head, and then I remembered that SteamVR now provide an OpenXR implementation (which seems to also be monado on Linux).

Long story short, I was able to run and test the official Godot VR example (the one with the sword, shotgun, grenades, ...) by combining SteamVR Beta (which provides an OpenXR library and implementation), Godot (through Steam) and this plugin.
I'm currently writing a small Steam guide about it, on the Godot section. I'll see if that gets traction.
